File: dart-wrapper/cardinal_sdk/android/src/main/kotlin/com/icure/cardinal_sdk/SubscriptionPlugin.kt
package com.icure.cardinal_sdk

import com.icure.cardinal.sdk.dart.subscription.EntitySubscription
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

object SubscriptionPlugin : MethodChannel.MethodCallHandler {

	@Suppress("UNCHECKED_CAST")
	override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
		if (
			!dispatch(
				methodName = call.method,
				parameters = (call.arguments as Map<String, String>?).orEmpty()
			) { success, errorCode, errorMessage, errorDetail ->
				if (errorCode != null)
					result.error(errorCode, errorMessage, errorDetail)
				else
					result.success(success)
			}
		) result.notImplemented()
	}

	fun dispatch(
		methodName: String,
		parameters: Map<String, String>,
		resultCallback: (
			String?,
			String?,
			String?,
			String?,
		) -> Unit,
	): Boolean = when(methodName) {
		"close" -> close(parameters, resultCallback)
		"getCloseReason" -> getCloseReason(parameters, resultCallback)
		"getEvent" -> getEvent(parameters, resultCallback)
		"waitForEvent" -> waitForEvent(parameters, resultCallback)
		else -> null
	}?.let { true } ?: false

	private fun close(parameters: Map<String, String>, resultCallback: (
		String?,
		String?,
		String?,
		String?,
	) -> Unit) {
		EntitySubscription.close(
			resultCallback,
			parameters.getValue("subscriptionId")
		)
	}

	private fun getCloseReason(parameters: Map<String, String>, resultCallback: (
		String?,
		String?,
		String?,
		String?,
	) -> Unit) {
		EntitySubscription.getCloseReason(
			resultCallback,
			parameters.getValue("subscriptionId")
		)
	}

	private fun getEvent(parameters: Map<String, String>, resultCallback: (
		String?,
		String?,
		String?,
		String?,
	) -> Unit) {
		EntitySubscription.getEvent(
			resultCallback,
			parameters.getValue("subscriptionId")
		)
	}

	private fun waitForEvent(parameters: Map<String, String>, resultCallback: (
		String?,
		String?,
		String?,
		String?,
	) -> Unit) {
		EntitySubscription.waitForEvent(
			resultCallback,
			parameters.getValue("cancellationToken").toLong(),
			parameters.getValue("subscriptionId"),
			parameters.getValue("timeout")
		)
	}

}

File: dart-wrapper/cardinal_sdk/ios/Classes/SubscriptionPlugin.swift
import Flutter
import CardinalDartSdkSupportLib

class SubscriptionPlugin {

    public static func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as! [String: String]
        let methodName = call.method

        let completed = SubscriptionPlugin.dispatch(methodName: methodName, parameters: args) { success, errorCode, errorMessage in
            if (errorCode != nil){
                result(FlutterError(code: errorCode!, message: errorMessage, details: nil))
            }
            else {
                result(success)
            }
        }

        if !completed {
            result(FlutterMethodNotImplemented)
        }
    }

  static func dispatch(
    methodName: String,
    parameters: [String : String],
    resultCallback: @escaping (
      String?,
      String?,
      String?
    ) -> Void
  ) -> Bool {
    switch methodName {
		case "close": close(parameters: parameters, resultCallback: resultCallback)
		case "getCloseReason": getCloseReason(parameters: parameters, resultCallback: resultCallback)
		case "getEvent": getEvent(parameters: parameters, resultCallback: resultCallback)
		case "waitForEvent": waitForEvent(parameters: parameters, resultCallback: resultCallback)
    	default: return false
    }
    return true
  }

	private static func close(
		parameters: [String : String],
		resultCallback: @escaping (String?,String?, String?) -> Void
	) {
		EntitySubscription.shared.close(
			dartResultCallback: resultCallback,
			subscriptionId: parameters["subscriptionId"]!
		)
	}

	private static func getCloseReason(
		parameters: [String : String],
		resultCallback: @escaping (String?,String?, String?) -> Void
	) {
		EntitySubscription.shared.getCloseReason(
			dartResultCallback: resultCallback,
			subscriptionId: parameters["subscriptionId"]!
		)
	}

	private static func getEvent(
		parameters: [String : String],
		resultCallback: @escaping (String?,String?, String?) -> Void
	) {
		EntitySubscription.shared.getEvent(
			dartResultCallback: resultCallback,
			subscriptionId: parameters["subscriptionId"]!
		)
	}

	private static func waitForEvent(
		parameters: [String : String],
		resultCallback: @escaping (String?,String?, String?) -> Void
	) {
		EntitySubscription.shared.waitForEvent(
			dartResultCallback: resultCallback,
			subscriptionId: parameters["subscriptionId"]!,
			timeoutString: parameters["timeout"]!
		)
	}
}
